Implement __builtin_volatile_load and __builtin_volatile_store

jpf91 opened the pull request:
Currently limited to isIntegral types. PODs could work as well, but the current implementation doesn't work with aggregate types (I guess the declarations/members have to be marked as volatile as well? It's quite a mystery to me when GCC uses what flag. For example TYPE_VOLATILE doesn't do anything without TREE_THIS_VOLATILE on the INDIRECT_REF).

Also loading a complete aggregate with such a builtin is probably almost always a performance problem (because of volatile guarantees you always have to load the complete struct, even if only one field is accessed. So it's better to expose a 'flat' API)

jpf91 committed:
Implement __builtin_volatile_load and __builtin_volatile_store

ibuclaw commented on gcc/d/d-codegen.cc:
```diff
@@ -2843,6 +2844,22 @@ expand_intrinsic (tree callexp)
+	case INTRINSIC_VOLATILE_LOAD:
+	  op1 = ce.nextArg();
+	  volatile_type = build_qualified_type (TREE_TYPE (callexp), TYPE_QUAL_VOLATILE);
+	  callexp = build1 (INDIRECT_REF, volatile_type, build1 (NOP_EXPR, build_pointer_type (volatile_type), op1));
```
`callexp = indirect_ref (volatile_type, op1);`

ibuclaw commented on gcc/d/d-codegen.cc:
```diff
@@ -2843,6 +2844,22 @@ expand_intrinsic (tree callexp)
+	  op1 = ce.nextArg();
+	  op2 = ce.nextArg();
+	  volatile_type = build_qualified_type (TREE_TYPE (op2), TYPE_QUAL_VOLATILE);
+	  callexp = build1 (INDIRECT_REF, volatile_type, build1 (NOP_EXPR, build_pointer_type (volatile_type), op1));
```
`callexp = indirect_ref (volatile_type, op1);`

ibuclaw commented on gcc/d/d-codegen.cc:
```diff
@@ -2843,6 +2844,22 @@ expand_intrinsic (tree callexp)
+	  volatile_type = build_qualified_type (TREE_TYPE (op2), TYPE_QUAL_VOLATILE);
+	  callexp = build1 (INDIRECT_REF, volatile_type, build1 (NOP_EXPR, build_pointer_type (volatile_type), op1));
+	  TREE_THIS_VOLATILE (callexp) = true;
+	  callexp = build2 (MODIFY_EXPR, volatile_type, callexp, op2);
```
`callexp = modify_expr (volatile_type, callexp, op2);`

ibuclaw commented on gcc/d/d-codegen.cc:
```diff
@@ -2843,6 +2844,22 @@ expand_intrinsic (tree callexp)
+	  op1 = ce.nextArg();
+	  op2 = ce.nextArg();
+	  volatile_type = build_qualified_type (TREE_TYPE (op2), TYPE_QUAL_VOLATILE);
+	  callexp = indirect_ref (volatile_type, build1 (NOP_EXPR, build_pointer_type (volatile_type), op1));
```
You still shouldn't need the NOP_EXPR here, `indirect_ref` takes care of that for you.

jpf91 replied:
Right, I've removed the NOP_EXPR.

ibuclaw commented on gcc/d/d-builtins.cc:
```diff
@@ -341,6 +342,81 @@ build_expression (tree cst)
+#include "d-intrinsics.def"
+#undef DEF_INTRINSIC
+  INTRINSIC_LAST
+};
```
Not sure about duplicating this enum here.

ibuclaw commented:
I keep thinking that perhaps we should wait until https://github.com/D-Programming-Language/druntime/pull/892 and have the intrinsic in `core.bitop`, or wherever it may go.

ibuclaw commented:
> For example TYPE_VOLATILE doesn't do anything without TREE_THIS_VOLATILE on the INDIRECT_REF

Yep, that is one reason why volatile statements were broken in D1 too. :)

jpf91 commented:
We can wait for D-Programming-Language/druntime#892 and hook directly into these intrinsics. I initially made this a template because I wanted to make it work with any POD type, but now I think it's not worth the effort and the core.bitop interface should be sufficient.

Anyway, I'm using this to test some Volatile!T implementations, but I'm not really happy so far ;-)
I see two approaches to implement it:

``` d
struct Volatile(T)
{
    T _store;
    opUnary...
}
enum TimerA =  cast(Volatile!(uint)*)0xABCD;

//But now the user always has to dereference:
(*TimerA)++;
(*TimerA) |= 0b1;
//Also unary operators don't work (bug?)
var has no effect in expression (__pitmp1352)
```

``` d
struct Volatile(T)
{
    T* _store;
    opUnary...
}
enum TimerA =  Volatile!(uint)(0xABCD);
TimerA++;
//Volatile(cast(uint*)3735928559LU) is not an lvalue
//But when not using enum the Volatile!T struct actually takes up space for every variable.
```

Maybe I'm missing something, but it looks as if we need a `pragma(address, )` or `@attribute(address, )`:

``` d
@attribute("address", 0xDEADBEEF) Volatile!uint TimerA;
TimerA++;
```

jpf91 committed:
Don't emit forceinlined functions

jpf91 committed:
Add @attribute("notypeinfo")

jpf91 committed:
Add @attribute("nocode")

jpf91 committed:
Add @attribute("noinit")

jpf91 committed:
Implement pragma(GNU_nomoduleinfo)

jpf91 commented:
@ibuclaw I added some changes which are not really related to peek/poke but I thought it's useful to add them to this pull request for review:
- Don't emit `forceinline`d functions (Need to unset weak & comdat for templates, cause functions can't be !TREE_PUBLIC and COMDAT or WEAK)
- Added some new attributes. These don't use the GCC attribute backend, they are only used in the frontend/glue layer. Added `getUDA(name, UserAttributeDeclaration)` to easily get the desired attribute.
  - Added `@attribute("notypeinfo")`: Don't emit typeinfo for class/struct
  - Added `@attribute("noinit")`: Don't emit initializer symbol `_initZ` for class/struct
  - Added `@attribute("nocode")`: Don't emit function (for CTFE only functions)
- Added `pragma(GNU_nomoduleinfo);`. Unfortunately can't be implemented with `@attribute` as it's not possible to attach UDAs to modules.

I also tried to implement gcc's `@attribute(optimize)`, then I realized that it's not useful in my case as it does not apply to inlined code. But if you're interested in this I can post the code. It was basically working fine.

So with this code:

``` d
//memory.d
module memory;
import gcc.builtins, gcc.attribute;

pragma(GNU_no_moduleinfo);
@attribute("noinit") @attribute("notypeinfo") struct Volatile(T) if (__traits(isIntegral, T))
{
    private T _store = void;

    @attribute("forceinline") void opOpAssign(string op)(in T rhs) nothrow
    {
        auto val = __builtin_volatile_load(&_store);
        mixin("val" ~ op ~ "= rhs;");
        __builtin_volatile_store(&_store, val);
    }
}

@attribute("nocode") Volatile!(T)* volatilePtr(T)(size_t addr)
{
    return cast(Volatile!(T)*)addr;
}
```

``` d
//main.d
import memory;
pragma(GNU_no_moduleinfo);
enum TimerB = volatilePtr!uint(0xDEADBEEF);

void test()
{
    *TimerB |= 0b1;
    *TimerB += 1;
}
```

we now generate:

``` asm
    .file   "memory.d"
    .section    .note.GNU-stack,"",@progbits
```

``` asm
; Without optimization
    .file   "main.d"
    .text
    .globl  _D4main4testFZv
    .type   _D4main4testFZv, @function
_D4main4testFZv:
.LFB0:
    .cfi_startproc
    pushq   %rbp
    .cfi_def_cfa_offset 16
    .cfi_offset 6, -16
    movq    %rsp, %rbp
    .cfi_def_cfa_register 6
    movl    $1, -8(%rbp)
    movl    $3735928559, %eax
    movl    (%rax), %eax
    movl    %eax, -12(%rbp)
    movl    -8(%rbp), %eax
    orl %eax, -12(%rbp)
    movl    $3735928559, %eax
    movl    -12(%rbp), %edx
    movl    %edx, (%rax)
    movl    $1, -4(%rbp)
    movl    $3735928559, %eax
    movl    (%rax), %eax
    movl    %eax, -16(%rbp)
    movl    -4(%rbp), %eax
    addl    %eax, -16(%rbp)
    movl    $3735928559, %eax
    movl    -16(%rbp), %edx
    movl    %edx, (%rax)
    nop
    popq    %rbp
    .cfi_def_cfa 7, 8
    ret
    .cfi_endproc
.LFE0:
    .size   _D4main4testFZv, .-_D4main4testFZv
    .ident  "GCC: (GNU) 4.10.0 20140629 (experimental)"
    .section    .note.GNU-stack,"",@progbits
```

``` asm
; With -O
    .file   "main.d"
    .text
    .globl  _D4main4testFZv
    .type   _D4main4testFZv, @function
_D4main4testFZv:
.LFB0:
    .cfi_startproc
    movl    $3735928559, %eax
    movl    (%rax), %edx
    orl $1, %edx
    movl    %edx, (%rax)
    movl    (%rax), %edx
    addl    $1, %edx
    movl    %edx, (%rax)
    ret
    .cfi_endproc
.LFE0:
    .size   _D4main4testFZv, .-_D4main4testFZv
    .ident  "GCC: (GNU) 4.10.0 20140629 (experimental)"
    .section    .note.GNU-stack,"",@progbits
```

The optimized code matches the C code. Unfortunately the unoptimized code is still two times bigger than in C. I'd like to fix this ultimately by using the frontend-inliner and adding some simple optimization to the frontend (this is really simple: compile time known `this` address and compile time known `rhs`) but the frontend-inliner needs to be fixed first. (where was this recent discussion on a dmd pull request about that again?)

@JinShil are these attributes useful for you as well? Do you have some other feedback? Do you think having to dereference when using `volatile!T` variables is OK?

ibuclaw commented:
I'll have a proper look later, but just a note that I've got it planned to do something like:

```
deprecated("Use __attribute__")
alias attribute = __attribute__;

auto __attribute__(A...)(A args) if(A.length > 0 && is(A[0] == string))
{
    return Attribute!A(args);
}

enum forceinline = __attribute__("forceinline");
enum noinline = __attribute__("noinline");
// etc...

auto target(A...)(A args) 
    if(A.length > 0 && is(A[0] == string))
{    
    return __attribute__("target", args);
} 
```

It's use being:

```
import gcc.attribute;

@forceinline void foo();
@target("sse3") void bar();
```

JinShil commented:
> @JinShil are these attributes useful for you as well? Do you have some other feedback?

In general, I prefer to avoid anything compiler-specific.  The idealist in me wants to try and implement as much as I can using only what's given in the language spec, but I'm a realist as well.  Sometimes, these features are the only way.

> `pragma(GNU_nomoduleinfo)`

I actually want to use `ModuleInfo` in my low level code because I want to be able to use the static constructors.  If I use `-fno-moduleinfo` then I don't get any `ModuleInfo` whatsoever, but with this feature it's possible to be selective on a type-by-type basis.  LDC already has something similar (http://wiki.dlang.org/LDC-specific_language_changes#LDC_no_moduleinfo)

> `@attribute("notypeinfo")`
> `@attribute("noinit")`

These are also quite nice, but are they necessary in the presence of `-ffunction-sections`, `-fdata-sections` and `-gc-sections`?  Or, are we trying to render `-gc-sections` obsolete?

> `@attribute("nocode")`
> Don't emit `forceinlined` functions

Again, I wonder if this should be handled by the linker.  Shouldn't linker optimizations strip these out?

I prefer an intelligent linker that can strip out unused inlined functions, `TypeInfo`, `_initZ`, etc... over decorating code with these attributes.  But in the absence of such a linker, I'll take it.

> Do you think having to dereference when using `Volatile!T` variables is OK?

There has to be a dereference somewhere.  `volatilePtr!(T)` makes perfect sense, but it should also be possible to emplace `Volatile!(T)` or create a templated type that encapsulates the dereference.  I'll probably do the latter in my code.

jpf91 commented:
> or create a templated type that encapsulates the dereference. I'll probably do the latter in my code.

I'd prefer that but I couldn't make it work with zero overhead:
-  `enum var = Volatile!(T,addr)()`: doesn't allow `|=` on enum literals, even if the type implements `opAssign` as there's no this pointer
- `var = Volatile!T(addr)` now requires additional space in data segment
- `@attribute(address) Volatile!T var;` is the only way I think this could work

Regarding other comments: Yes, the linker can do all that as well. I'd still prefer this to be done in the compiler: Less work for the linker, no need to use these linker flags. Of course the idea is not to mark every struct with `attribute("notypeinfo")`, there should be a compiler switch for this. The attributes are just for code were you might want typeinfo in some cases, but you know for sure that you don't want it for `Volatile!T`. Some of these attributes could be moved to the frontend later on (notypeinfo) and not emitting force-inlined functions is a logical optimization for forceinline (if a function is always inlined, there's no way to call it, so there's no need to output it).

@ibuclaw I saw that post in the newsgroup. I think this will work just fine with these new attributes.

ibuclaw commented:
I ended up dropping off the conversation and forgetting what was going on.  I take it peek/poke never happened?   This is a shame.

jpf91 commented:
IIRC it was merged into dmd 2.067. I just closed outdated pull requests but I'll probably have to revisit this for 2.067.
